### ISSUE TITLE
Feature request : i18n support for emails and application settings

### ISSUE THREAD

**Issue by ymauray** [pocketbase/pocketbase#593]
Hi, thanks for the great tool !

I am building a Flutter mobile app that will support english and french as its languages. It would be very nice to have the possibiliy to send localized version of the `Verification`, `Password reset`, and `Confirm email change` emails. Manual selection of the language via the API call is perfectly acceptable. It could be english by default. 

Also, the name of the app will be localized, so it won't be the same in english and in french. That could be hard coded in the email templates, so having that localized in PocketBase is not as important as the emails templates.

Have a good day !

**Comment by ganigeorgiev** [comment]
I've recently received a similar request on the support mail, but unfortunately this cannot be done easily at the moment. I'll leave the issue open for now, so that I can consider it and maybe combine with the planned collection triggers in https://github.com/pocketbase/pocketbase/issues/317.

---

Currently there are 2 available options if you want to have localized mail templates:

1. Extend PocketBase and use the [`OnMailerBefore*` event hooks](https://pocketbase.io/docs/event-hooks/#mailer-hooks) to send custom emails (you have access to the http context and can read the user language from a query parameter or the `Accept-Language` header).

2. Provide both localized versions in the same mail template. I've seen this pattern used with Firebase, and although it is not ideal, it maybe a good enough workaround for now, especially for users who are not familiar with Go and don't want to use PocketBase as framework.

**Comment by ymauray** [comment]
Thanks for the suggestions. I totally understand that it might be complex - i18n things are always complex ! - and it's more important to make PocketBase work well first ! 

I'll figure something out for the name of my app, and I like the idea of the emails having both languages, so I'll go with that until it's possible to have them localized. 

Cheers from Switzerland!

**Comment by ganigeorgiev** [comment]
@gedw99 There is no external template language if that's what you are asking. Internally we are using the go `text/template` and `html/template` packages.

For general info how to send emails when using PocketBase as framework, you can check https://pocketbase.io/docs/sending-emails/

**Comment by gladson** [comment]
To help with the discussion...

https://github.com/maximilien/i18n4go

**Comment by ameerkabir** [comment]
I am also interested in building with Pocketbase but the blocker is internationalisation. I am happy to crack on given enough guidance on this.

**Comment by flerovium11** [comment]
Had the same issue and ended up building a workaround using Extend PocketBase with mailer event hooks.

Full documentation:
https://medium.com/@contact_48948/localizing-pocketbase-emails-with-javascript-hooks-30a6f0b7ad5a

### TLDR;
- Create HTML email templates per locale in `/pb_hooks/templates/{locale}/`
- Load and cache templates on `onBootstrap`
- Detect locale from Accept-Language or a custom header (X-Locale) passed from the frontend
- Store locale in `$app.store()` together with `record.id`
- Override PocketBase mailer events (onMailerRecordVerificationSend, onMailerRecordEmailChangeSend)
- Inject dynamic values (eg. verification links) into templates

Frontend sets Accept-Language header:
```
await pb.collection('users').requestVerification(email, { 
  headers: { 'Accept-Language': locale } 
})
```

localize_emails.pb.js
```javascript
/// <reference path="../pb_data/types.d.ts" />

onBootstrap((e) => {
  const utils = require(`${__hooks}/utils.js`)
  utils.loadTemplates()
  console.log('Email localization hooks initialized')
  e.next()
})

onMailerRecordVerificationSend((e) => {
  const utils = require(`${__hooks}/utils.js`)
  const locale = utils.getLocale(e.record)
  const { subject, html } = utils.loadTemplate(locale, 'verification')

  e.message.subject = subject
  const token = e.meta.token
  if (!token) throw new Error('Verification token not found in event metadata')
  const appUrl = $app.settings().meta.appURL
  if (!appUrl) throw new Error('App URL not found in app settings')
  const verificationLink = `${appUrl}/auth/confirm-verification?token=${token}`
  utils.applyTemplate(e.message, html, { VERIFICATION_LINK: verificationLink })

  e.next()
})

onMailerRecordEmailChangeSend((e) => {
  const utils = require(`${__hooks}/utils.js`)
  const locale = utils.getLocale(e.record)
  const { subject, html } = utils.loadTemplate(locale, 'email_change')

  e.message.subject = subject
  const token = e.meta.token
  if (!token) throw new Error('Email change token not found in event metadata')
  const appUrl = $app.settings().meta.appURL
  if (!appUrl) throw new Error('App URL not found in app settings')
  utils.applyTemplate(e.message, html, { TOKEN: token })

  e.next()
})

// Capture locale on requests that will subsequently trigger an email

onRecordRequestVerificationRequest((e) => {
  const utils = require(`${__hooks}/utils.js`)
  utils.captureLocale(e)
  e.next()
})

onRecordRequestEmailChangeRequest((e) => {
  const utils = require(`${__hooks}/utils.js`)
  utils.captureLocale(e)
  e.next()
})
```
utils.js
```javascript
const supportedLocales = ['en', 'de']
const fallbackLocale = 'en'

// Call this from onBootstrap
function loadTemplates() {
  // Can be extended with 'password_reset', 'otp', 'login_alert'
  const templateTypes = ['verification', 'email_change']
  const styles = toString($os.readFile(`${__hooks}/templates/_styles.css`))

  const templates = {}
  for (const locale of supportedLocales) {
    for (const type of templateTypes) {
      const key = `${locale}/${type}`

      // toString is a helper provided by the environment
      // TextEncoder is not available here
      // String.fromCharCode(...) does not support multi-byte characters
      let html = toString($os.readFile(`${__hooks}/templates/${locale}/${type}.html`))
      html = html.replace('{STYLES}', `<style>${styles}</style>`)
      const match = html.match(/<!--\s*SUBJECT:\s*(.+?)\s*-->/)
      if (!match) {
        throw new Error(`Subject not found in template ${key}`)
      }
      html = html.replace(/<!--\s*SUBJECT:\s*.+?-->\n?/, '')
      templates[key] = { subject: match[1].trim(), html }
    }
  }
  $app.store().set('emailTemplates', templates)
}

// Called by request hooks that trigger an email send
function captureLocale(e) {
  const raw =
    e.requestEvent.request.header.get('Accept-Language') ||
    e.requestEvent.request.header.get('X-Locale') ||
    fallbackLocale
  const locale = raw.split(/[-_]/)[0].toLowerCase()
  if (supportedLocales.includes(locale)) $app.store().set(`pendingLocales__${e.record.id}`, locale)
}

// Get locale from user record and fallback to locale
// stored by captureLocale on previous request from same user
function getLocale(record) {
  const stored = record.get('locale')
  if (stored && supportedLocales.includes(stored)) return stored
  const key = `pendingLocales__${record.id}`
  const pending = $app.store().get(key)
  if (pending) {
    $app.store().remove(key)
    return pending
  }
  return fallbackLocale
}

function loadTemplate(locale, type) {
  if (!supportedLocales.includes(locale)) locale = fallbackLocale

  const templates = $app.store().get('emailTemplates')
  const key = `${locale}/${type}`
  if (!templates[key]) {
    throw new Error(`Template not found for locale "${locale}" and type "${type}"`)
  }

  return templates[key]
}

function applyTemplate(message, html, placeholders) {
  let body = html
  for (const [key, value] of Object.entries(placeholders)) {
    body = body.split(`{${key}}`).join(value)
  }
  message.html = body
}

module.exports = {
  loadTemplates,
  applyTemplate,
  captureLocale,
  getLocale,
  loadTemplate,
}
```
Happy to adapt as soon as there's an official solution.
